sklearn/preprocessing/_encoders.py

```diff
@@ -38,44 +38,65 @@ def _check_X(self, X):
         - convert list of strings to object dtype
         - check for missing values for object dtype data (check_array does
           not do that)
+        - return list of features (arrays)
 
         """
-        X_temp = check_array(X, dtype=None)
-        if not hasattr(X, 'dtype') and np.issubdtype(X_temp.dtype, np.str_):
-            X = check_array(X, dtype=np.object)
+
+        if not (hasattr(X, 'iloc') and getattr(X, 'ndim', 0) == 2):
+            X_temp = check_array(X, dtype=None)
+            if not hasattr(X, 'dtype') and\
+               np.issubdtype(X_temp.dtype, np.str_):
+                X = check_array(X, dtype=np.object)
+            else:
+                X = X_temp
+            needs_validation = False
         else:
-            X = X_temp
+            needs_validation = True
 
-        if X.dtype == np.dtype('object'):
-            if not _get_config()['assume_finite']:
-                if _object_dtype_isnan(X).any():
-                    raise ValueError("Input contains NaN")
+        n_samples, n_features = np.shape(X)
```

Comment:
you can use `X.shape` here instead (since both arrays and dataframes have this attribute)

Reply:
Thanks, I added the comments